### PR TITLE
Automated Name Recommendation For The Extract Local Variable Refactoring

### LINKED ISSUE #601: Automated Name Recommendation For The Extract Local Variable Refactoring
### Automated Name Recommendation For The Extract Local Variable Refactoring
---
**Description**
---
"Extract Variable" refactoring is a useful feature of JDT. However, the names recommended for the extracted variables sometimes are substantially different from what developers manually coin. Here we report a case as follows:
This example is a real-world case of extracting local variable refactoring in GeoServer, an open-source software server written in Java that allows users to share and edit geospatial data. (https://github.com/geoserver/geoserver/commit/c54d9ccbebd5d55ff2bcb2ca4db5e7167c1e2f2a#diff-40dd1af56b59d542f9c41036b47c0db77b2b4a5bc8ece7a2f4a78cc196867181, line 61 in src/community/oseo/oseo-stac/src/main/java/org/geoserver/ogcapi/stac/TemplatedItemsConverter.java )
```java
/* CS1: Real Case of Extract Local Variable Refactoring */
protected void writeInternal(
            AbstractItemsResponse itemsResponse, HttpOutputMessage httpOutputMessage)
            throws IOException {
-		RootBuilder builder = templates.getItemTemplate();
        try (STACGeoJSONWriter writer =
                new STACGeoJSONWriter(
                        new JsonFactory()
                                .createGenerator(httpOutputMessage.getBody(), JsonEncoding.UTF8))) {
            writer.startTemplateOutput(null);
            try (FeatureIterator features = itemsResponse.getItems().features()) {
                while (features.hasNext()) {
-					builder.evaluate(writer, new TemplateBuilderContext(features.next()));
+       			// lookup the builder, might be specific to the parent collection
+        			Feature feature = features.next();
+        			String collectionId =
+                		(String) feature.getProperty("parentIdentifier").getValue();
+        			RootBuilder builder = templates.getItemTemplate(collectionId);
+        			builder.evaluate(writer, new TemplateBuilderContext(feature));
                }
            }
            //Do other things
      	    // ...
        } catch (Exception e) {
            throw new ServiceException(e);
        }
    }
```
The actual name developer assigned to the extracted expression, i.e., `features.next()` (line 13 in CS1) is "feature". However, if we select the expression features.next() and conduct refactoring "extract local variable" in Eclipse, we will get the following code snippet:
```java
/* CS2: Code Snippet After Refactoring in Eclipse*/
protected void writeInternal(
            AbstractItemsResponse itemsResponse, HttpOutputMessage httpOutputMessage)
            throws IOException {
		RootBuilder builder = templates.getItemTemplate();
        try (STACGeoJSONWriter writer =
                new STACGeoJSONWriter(
                        new JsonFactory()
                                .createGenerator(httpOutputMessage.getBody(), JsonEncoding.UTF8))) {
            writer.startTemplateOutput(null);
            try (FeatureIterator features = itemsResponse.getItems().features()) {
                while (features.hasNext()) {
					Feature next = features.next();
					builder.evaluate(writer, new TemplateBuilderContext(feature));
                }
            }
            //Do other things
      	    // ...
        } catch (Exception e) {
            throw new ServiceException(e);
        }
    }
```
The name recommended for the extracted variable is "next" which is substantially different from what the developers finally selected. 

**Solution**
---
To this end, we proposed a new rule for recommending names for extracted variables:
`X = Xs.next()`, where "Xs" represents the receiver object in plural format, and "X" represents the singular format of the receiver object. Note that there are multiple plural forms, e.g., *f -> *ves, and here we just take "X" and "Xs" for example. 

The precondition for applying this rule is specified as follows:
1. The expression to be extracted is an instance of MethodInvocation.
2. The selected expression contains a receiver that is a SimpleName.
3. The receiver object is in plural format, indicating that it is a collection object.
4. The invoked method is named "next", indicating that the expression is to retrieve a single element from the collection.

To validate the effectiveness of the proposed rule, we identified real-world  extracted variables refactorings by mining commit histories of projects on GitHub.  
Evaluation results suggest that Xs.next() is often extracted as local variables, and developers rarely (2 out of 26 cases) name the variables as "next" (as Eclipse recommends). 
In contrast, the manually selected variable names are identical to the ones recommended by the new rule in 15 out of 26 cases, much higher than the success rate of Eclipse's default recommendation. 

**We will submit the pull request (implementing the new rule) if that is fine.**

### PR DESCRIPTION
**Pull request by Michaelll123**
- modify StubUtility() to add new rules to generate names for extracted local variables.
- add popular plural words (irregular cases) to ConvertLoopOperation.java to support the singularization of plural words.
- add test to ExtractTempTests.
- fixes #601 .

## What it does
see issue cited.

## How to test
Run test153, and test154 in ExtractTempTests.
## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java [hunk 1]
```diff
@@ -0,0 +1,295 @@
+package org.eclipse.jdt.internal.core.manipulation;
+/**
+ * This utility class for English words is taken from a blog on CSDN, and we modify it to adapt for the cases of multiply word.
+ * The original blog can be found in https://blog.csdn.net/weixin_44258855/article/details/123201283
```
Comment (iloveeclipse) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java:4]: Sorry, this can't be merged into eclipse.org codebase as it refers to a source without clear licensing information, doesnt represent your own code and the snippet itself seem to be largely inadequate to address the problem you are trying to fix.

**Review by iloveeclipse — CHANGES_REQUESTED** [review]

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java (hunk 1 above)
Comment (Michaelll123) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java:4]: Thanks, I will try to solve this problem soon.

Comment (aino-gautam) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java:4]: @iloveeclipse how did you determine the source of code ? i.e if its devs own or not ? Is there some tool for it ?

**Comment by Michaelll123** [comment]
Hi, I am a green hand in contributing to eclipse. In the commits earlier I am trying to locate the errors of my code. &nbsp;I am trying to implement my own version of this utility class (org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java). When I finished, I will submit a new pull request.&nbsp;Thanks a lot for your precious time.

发自我的iPhone


------------------ Original ------------------
From: Debasish padhy ***@***.***&gt;
Date: Wed,May 31,2023 11:34 AM
To: eclipse-jdt/eclipse.jdt.ui ***@***.***&gt;
Cc: Michaelll123 ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [eclipse-jdt/eclipse.jdt.ui] Automated Name Recommendation ForThe Extract Local Variable Refactoring (PR #602)





 
@aino-gautam commented on this pull request.
 
 
In org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java:
 &gt; @@ -0,0 +1,295 @@ +package org.eclipse.jdt.internal.core.manipulation; +/** + * This utility class for English words is taken from a blog on CSDN, and we modify it to adapt for the cases of multiply word. + * The original blog can be found in https://blog.csdn.net/weixin_44258855/article/details/123201283  
@iloveeclipse how did you determine the source of code ? i.e if its devs own or not ? Is there some tool for it ?

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java (hunk 1 above)
Comment (AnonymousAccount4SE) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java:4]: Are third-party APIs allowed to merge into eclipse.org codebase?

Comment (iloveeclipse) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java:4]: > Are third-party APIs allowed to merge into eclipse.org codebase?

See https://www.eclipse.org/projects/handbook/#ip

TL;DR: content added to the jdt.ui bundle must be licensed under EPL license.

Comment (iloveeclipse) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java:4]: > @iloveeclipse how did you determine the source of code ? i.e if its devs own or not ? Is there some tool for it ?

I would name it BDA  "brain driven analysis". Of course as every tool that might fail too.
In this concrete context I just assume the "using code posted by someone on some blog page" is most likely description of the code not written by PR author and so it is highly questionable if that is allowed to be merged here.

Comment (Michaelll123) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java:4]: It looks like there is no existing open-source tool written in Java to support the convertion between plural words and singular words.  I will implement my own version soon.

Comment (jjohnstn) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java:4]: @Michaelll123 FWIW: a lot of English words are unnecessary to handle (e.g. geese to goose is unlikely to show up in code and not worth the time to look for it).  There was a pared down version of finding a singular in ConvertLoopOperation.modifyBaseName() which gets a variable name for an enhanced-for-loop (e.g. for (String name : names)).  It handled a number of common usages (e.g. children vs child) and then identified odd English plural suffixes to avoid attempting removing the "s" in which case it defaulted to "element".   You could base off of that and add more direct conversion cases if you want.  Pure plural support is over-kill IMO.

Comment (Michaelll123) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java:4]: @jjohnstn Thanks for the insightful advice! It is indeed a tricky problem to cover all the conversion cases in English, and I will start with ConvertLoopOperation.modifyBaseName().

Comment (Michaelll123) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/EnglishWordUtil.java]: Thanks for all your valuable advice, and I have finished the requested changes. Please kindly review the latest pull request.

**Review by jjohnstn — CHANGES_REQUESTED** [review]
Overall, the logic is fine and additions to the loop name logic are useful.  The logic can be moved into another method in StubUtility and cleaned up a bit.

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/ConvertLoopOperation.java [hunk 2]
```diff
@@ -59,7 +60,21 @@ public abstract class ConvertLoopOperation extends CompilationUnitRewriteOperati
 			new AbstractMap.SimpleImmutableEntry<>("Children", "Child"), //$NON-NLS-1$ //$NON-NLS-2$
 			new AbstractMap.SimpleImmutableEntry<>("Entries", "Entry"), //$NON-NLS-1$ //$NON-NLS-2$
 			new AbstractMap.SimpleImmutableEntry<>("Proxies", "Proxy"), //$NON-NLS-1$ //$NON-NLS-2$
-			new AbstractMap.SimpleImmutableEntry<>("Indices", "Index")) //$NON-NLS-1$ //$NON-NLS-2$
+			new AbstractMap.SimpleImmutableEntry<>("Indices", "Index"), //$NON-NLS-1$ //$NON-NLS-2$
+			new AbstractMap.SimpleImmutableEntry<>("People", "Person"), //$NON-NLS-1$ //$NON-NLS-2$
+			new AbstractMap.SimpleImmutableEntry<>("Properties", "Property"), //$NON-NLS-1$ //$NON-NLS-2$
+			new AbstractMap.SimpleImmutableEntry<>("Factories", "Factory"), //$NON-NLS-1$ //$NON-NLS-2$
+			new AbstractMap.SimpleImmutableEntry<>("Archives", "archive"), //$NON-NLS-1$ //$NON-NLS-2$
```
[the remark this answers is not in the document]
Comment (Michaelll123) [reply org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/ConvertLoopOperation.java:67]: Thanks for the review. These cases indeed can be formed by removing the trailing "s", however, If they are not in  IRREG_NONS list, this method will first check if the names end with "ves"/"ces" (defined in IRREG_ENDINGS). If it is, default element is returned, leading to the "remove s" part not being executed. So I propose to keep these elements in this list.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/StubUtility.java [hunk 3]
```diff
@@ -1014,7 +1016,21 @@ private static String removeTypeArguments(String baseName) {
 	public static String[] getVariableNameSuggestions(int variableKind, IJavaProject project, ITypeBinding expectedType, Expression assignedExpression, Collection<String> excluded) {
 		LinkedHashSet<String> res= new LinkedHashSet<>(); // avoid duplicates but keep order
 
-		if (assignedExpression != null) {
+			if (assignedExpression != null) {
```
[the remark this answers is not in the document]
Comment (Michaelll123) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/StubUtility.java:1019]: Thanks for the review, I have done the formatting using CTRL + SHIFT + F.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/StubUtility.java [hunk 4]
```diff
@@ -1014,7 +1016,21 @@ private static String removeTypeArguments(String baseName) {
...
+				if(assignedExpression instanceof MethodInvocation) {
+					MethodInvocation methodInvocation= (MethodInvocation)assignedExpression;
+					Expression receiver= methodInvocation.getExpression();
+					String methodName= methodInvocation.getName().toString();
+					if(receiver != null) {
```
[the remark this answers is not in the document]
Comment (Michaelll123) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/StubUtility.java:1024]: Thanks for the review, and I have put my code snippets into getBaseNameFromExpression.

**Comment by Michaelll123** [comment]
Thanks for all your valuable advice, and I have finished the requested changes. Please kindly review the latest pull request.

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/ConvertLoopOperation.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jjohnstn) [reply org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/ConvertLoopOperation.java:67]: Ok.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/StubUtility.java [hunk 5]
```diff
@@ -1146,23 +1148,30 @@ private static void add(String[] names, Set<String> result) {
...
+			name= methodInvocation.getName().getIdentifier();
+			if (name.equals("next")) { //$NON-NLS-1$
+				Expression receiver= methodInvocation.getExpression();
+				String modifiedName= getBaseNameFromReceiver(receiver);
+				if (!modifiedName.equals("element")) //$NON-NLS-1$
+					return modifiedName;
+			}
 		} else if (assignedExpression instanceof SuperMethodInvocation) {
-			name= ((SuperMethodInvocation)assignedExpression).getName().getIdentifier();
+			name= ((SuperMethodInvocation) assignedExpression).getName().getIdentifier();
```
Comment (jjohnstn) [inline org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/StubUtility.java:1170]: Something in your format profile is adding a blank between casts and their element to cast.  This should not be changed as part of this patch.  Other than that, the code looks good.  If you want me to fix it with my own formatting, I can do so.

**Review by jjohnstn — CHANGES_REQUESTED** [review]
Looks good.  Just a formatting issue with casting adding a blank.  Fix that and it is ready to go.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/StubUtility.java (hunk 5 above)
Comment (Michaelll123) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/StubUtility.java:1170]: Thanks a lot! I will fix that immediately.

File: org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/StubUtility.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jjohnstn) [reply org.eclipse.jdt.core.manipulation/common/org/eclipse/jdt/internal/core/manipulation/StubUtility.java]: Looks good.

**Review by jjohnstn — CHANGES_REQUESTED** [review]
Hi @Michaelll123 The patch is fine but please recreate the patch in your branch so it is rebased on master and submit via --force push.  You have been merging master into your branch which is disallowed.  You should always rebase your patches on top of master as opposed to merging master into your branch.

**Comment by Michaelll123** [comment]
I am new to the process of committing, and I push an empty commit, leading to the pull request being closed. I am really sorry about that. Could you please reopen this pull request?

**Comment by Michaelll123** [comment]
Thanks for the review. I think the latest patch is based on master, and there are no commits merging master into recommend-name-for-extract-local-variables. Please check again, thanks!

**Review by jjohnstn — APPROVED** [review]

**Comment by jjohnstn** [comment]
Hi @Michaelll123 Looks good.  Just need to wait until master opens for 4.29 M1

**Comment by Michaelll123** [comment]
Thanks! Have a good one!
